packages/protobuf/src/private/names.ts

```diff
@@ -180,98 +168,144 @@ function protoCamelCase(snakeCase: string): string {
-  package: true,
-  private: true,
-  protected: true,
-  public: true,
-  static: true,
+  "enum",
+  "implements",
+  "interface",
+  "let",
+  "package",
+  "private",
+  "protected",
+  "public",
+  "static",
 
   // Class name cannot be 'Object' when targeting ES5 with module CommonJS
-  Object: true,
+  "Object",
 
   // TypeScript keywords that cannot be used for types (as opposed to variables)
-  bigint: true,
-  number: true,
-  boolean: true,
-  string: true,
-  object: true,
+  "bigint",
+  "number",
+  "boolean",
+  "string",
+  "object",
 
   // Identifiers reserved for the runtime, so we can generate legible code
-  globalThis: true,
-  Uint8Array: true,
-  Partial: true,
-};
+  "globalThis",
+  "Uint8Array",
+  "Partial",
+]);
 
-// Names that cannot be used for object properties because they are reserved
-// by built-in JavaScript properties.
-const reservedObjectProperties: { [k: string]: boolean } = {
+/**
+ * Names that cannot be used for object properties because they are reserved
+ * by built-in JavaScript properties.
+ */
+const reservedObjectProperties = new Set([
   // names reserved by JavaScript
-  constructor: true,
-  toString: true,
-  toJSON: true,
-  valueOf: true,
-};
+  "constructor",
+  "toString",
+  "toJSON",
+  "valueOf",
+]);
 
-// Names that cannot be used for object properties because they are reserved
-// by the runtime.
-const reservedMessageProperties: { [k: string]: boolean } = {
+/**
+ * Names that cannot be used for object properties because they are reserved
+ * by the runtime.
+ */
+const reservedMessageProperties = new Set([
   // names reserved by the runtime
-  getType: true,
-  clone: true,
-  equals: true,
-  fromBinary: true,
-  fromJson: true,
-  fromJsonString: true,
-  toBinary: true,
-  toJson: true,
-  toJsonString: true,
+  "getType",
+  "clone",
+  "equals",
+  "fromBinary",
+  "fromJson",
+  "fromJsonString",
+  "toBinary",
+  "toJson",
+  "toJsonString",
 
   // names reserved by the runtime for the future
-  toObject: true,
+  "toObject",
+]);
+
+const fallback = <T extends string>(name: T) => `${name}$` as const;
+
+/**
+ * Will wrap names that are Object prototype properties or names reserved
+ * for `Message`s.
+ */
+const safeMessageProperty = (name: string): string => {
+  if (
+    reservedObjectProperties.has(name) ||
+    reservedMessageProperties.has(name)
+  ) {
+    fallback(name);
+  }
+  return name;
+}
+
+/**
+ * Names that cannot be used for object properties because they are reserved
+ * by built-in JavaScript properties.
+ */
+export const safeObjectProperty = (name: string): string => {
+  if (reservedObjectProperties.has(name)) {
+    return fallback(name);
+  }
+  return name;
+};
+
+/**
+ * Names that can be used for identifiers or class properties
+ */
+export const safeIdentifier = (name: string): string => {
+  if (
+    reservedObjectProperties.has(name) ||
+    reservedIdentifiers.has(name)
+  ) {
+    return fallback(name);
+  }
+  return name;
```

Comment:
```suggestion
export const safeIdentifier = (name: string): string => {
  if (reservedIdentifiers.has(name)) {
    return fallback(name);
  }
  return name;
```

With this change, and the one above, all tests pass for me. Can you try that? Please make sure to run `make` and commit all diffs.